### PR TITLE
[FEAT] date and timestamp parsers

### LINKED ISSUE #2302: String column to datetime conversion
#2294

### LINKED ISSUE #2309: String column to date conversion
`str.to_date(format_str : str)`

An expression to convert strings to dates using a given format string, similar to https://ibis-project.org/reference/expression-strings#ibis.expr.types.strings.StringValue.to_date

See also: https://github.com/Eventual-Inc/Daft/issues/2302

### PR DESCRIPTION
**Pull request by murex971**
Resolves #2309
Resolves #2302

### CODE REVIEW

**Comment by samster25** [comment]
@kevinzwang can you take a look?

File: daft/expressions/expressions.py [hunk 1]
```diff
@@ -1436,6 +1436,30 @@ def substr(self, start: int | Expression, length: int | Expression | None = None
         length_expr = Expression._to_expression(length)
         return Expression._from_pyexpr(self._expr.utf8_substr(start_expr._expr, length_expr._expr))
 
+    def to_date(self, format: str | Expression) -> Expression:
```
Comment (colin-ho) [inline daft/expressions/expressions.py:1439]: Lets change this to just accept `str`. See the comment below for changing `to_datetime` to accept a single `str` as well.

File: daft/expressions/expressions.py [hunk 2]
```diff
@@ -1436,6 +1436,30 @@ def substr(self, start: int | Expression, length: int | Expression | None = None
...
+        Example:
+            >>> col("x").str.to_date("%Y-%m-%d")
+
+        Returns:
+            Expression: a Date expression which is parsed by given format
+        """
+        format_expr = Expression._to_expression(format)
+        return Expression._from_pyexpr(self._expr.utf8_to_date(format_expr._expr))
+
+    def to_datetime(self, format: str | Expression) -> Expression:
```
Comment (colin-ho) [inline daft/expressions/expressions.py:1451]: We need to support different timeunits, as well as datetimes with timezones. To do this, lets change the function signature to `to_datetime(self, format : str, timezone: str | None = None)`

By constraining format to be a single str, we can infer the correct timeunit when we call `to_field` in src/daft-dsl/src/functions/utf8/to_datetime.rs by checking if the format string contains nanosecond, i.e. `%9f` or `%.9f`, or millisecond, i.e. `%.3f` or `%3f`, precision, otherwise default to microsecond.

By allowing the `timezone: str | None = None` argument, we can parse the string using `chrono::NaiveDateTime::parse_from_str` or `chrono::DateTime::parse_from_str` at the array level.

File: tests/series/test_utf8_ops.py [hunk 3]
```diff
@@ -1270,3 +1272,205 @@ def test_series_utf8_substr_bad_dtype() -> None:
...
+        s.str.to_date(formats)
+
+
+def test_series_utf8_bad_date() -> None:
+    s = Series.from_arrow(pa.array(["2021-100-20"]))
+    formats = Series.from_arrow(pa.array(["%Y-%m-%d"]))
+    with pytest.raises(ValueError):
+        s.str.to_date(formats)
+
+
```
Comment (colin-ho) [inline tests/series/test_utf8_ops.py:1371]: Let's add tests for datetimes with timezones as well. A particular edge case to consider would be columns with different timezones, where each entry should be converted to the user provided timezone.

**Review by colin-ho — COMMENTED** [review]
Nice work so far. Overall, `to_date` looks good, and `to_datetime` can be enhanced to support timeunit inference and timezones

**Comment by samster25** [comment]
@murex971 Is this ready for rereview?

**Comment by colin-ho** [comment]
Thanks for this @murex971 ! I just made some additional commits to get this to the finish line, didn't think another round of back and forth was needed :)
